### ISSUE TITLE
Stack overflow caused by function recursion

### ISSUE THREAD

**Issue by rfiischer** [serge-sans-paille/pythran#1442]
I got an "fatal Python error: Cannot recover from stack overflow" and, after analyzing the printed stack, I found what I think is the cause of the error. I can't post the script that caused the error :(, but I can point it out.

```python
# From pythran/cxxtypes.py line 227
def generate(self, ctx):
    try:
        return 'typename __combined<{}>::type'.format(
            ','.join(ctx(t) for t in self.types))
    except RuntimeError:
        # this is a situation where we accept to somehow extend
        # the recursion limit, because of degenerated trees
        import sys
        current_recursion_limit = sys.getrecursionlimit()
        try:
            return self.generate(ctx)
        # This except is never executed
        except RuntimeError:
            sys.setrecursionlimit(current_recursion_limit * 2)
            return self.generate(ctx)
        finally:
            sys.setrecursionlimit(current_recursion_limit)
```

After analyzing the code one can realize that the pointed `except` is never executed. That's because if a `RuntimeError` is caused, it is caught by the second call of `self.generate`, and a infinite loop is generated.

I circumvented the error with the following substitution
```python
def generate(self, ctx):
    try:
        return 'typename __combined<{}>::type'.format(
            ','.join(ctx(t) for t in self.types))
    except RuntimeError:
        # this is a situation where we accept to somehow extend
        # the recursion limit, because of degenerated trees
        import sys
        current_recursion_limit = sys.getrecursionlimit()
        sys.setrecursionlimit(current_recursion_limit * 2)
        result = self.generate(ctx)
        sys.setrecursionlimit(current_recursion_limit)
        return result
```

It is not the ideal fix, because it assumes that just doubling the recursion limit will eventually cause the `try` statement to work.

**Comment by serge-sans-paille** [comment]
Your patch 100% makes sense, because of the recursion on self.generate. Do you want to issue a PR, or you'd like me to do so?

**Comment by rfiischer** [comment]
Yes I can do that! I just couldn't debug the code; if I put a breakpoint inside the except the debugger would crash. My doubt is whether this fix actually sets the recursion limit back to the original before the first call. I can issue a PR with the current fix though.

**Comment by serge-sans-paille** [comment]
Closed by #1472 and #1452
